File: leetcode.com/python/173_Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Approach 1: https://tinyurl.com/smu9ku3
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.nodesSorted = []
        self.nextIndex = 0
        self.__inorder(root)

    def __inorder(self, root):
        if not root:
            return
        self.__inorder(root.left)
        self.nodesSorted.append(root.val)
        self.__inorder(root.right)

    def next(self):
        """
        @return the next smallest number
        :rtype: int
        """
        value = self.nodesSorted[self.nextIndex]
        self.nextIndex += 1
        return value

    def hasNext(self):
        """
        @return whether we have a next smallest number
        :rtype: bool
        """
        return self.nextIndex < len(self.nodesSorted)


# Approach 2: https://tinyurl.com/smu9ku3
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        self.__leftMostInorder(root)

    def __leftMostInorder(self, root):
        curentStack = self.stack
        while root:
            curentStack.append(root)
            root = root.left

    def next(self):
        """
        @return the next smallest number
        :rtype: int
        """
        curentStack = self.stack
        topMostNode = curentStack.pop()
        if topMostNode.right:
            self.__leftMostInorder(topMostNode.right)
        return topMostNode.val

    def hasNext(self):
        """
        @return whether we have a next smallest number
        :rtype: bool
        """
        curentStack = self.stack
        return len(curentStack) > 0

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/173_Binary_Search_Tree_Iterator.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */

class BSTIterator {

    var nodeStack = [TreeNode]()
    
    init(_ root: TreeNode?) {
        var currentNode: TreeNode? = root
        while var node = currentNode  {
            nodeStack.append(node)
            currentNode = node.left
        }
    }
    
    /** @return the next smallest number */
    func next() -> Int {
        if var currentNode = nodeStack.popLast() {
            let valToReturn = currentNode.val
            if let right = currentNode.right {
                nodeStack.append(right)
                currentNode = right
                while let node = currentNode.left  {
                    nodeStack.append(node)
                    currentNode = node
                }
            }
            return valToReturn
        } else {
            return -1
        }
    }
    
    /** @return whether we have a next smallest number */
    func hasNext() -> Bool {
        return nodeStack.count > 0 ? true : false
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator(root)
 * let ret_1: Int = obj.next()
 * let ret_2: Bool = obj.hasNext()
 */
